### PR TITLE
DHFPROD-6172: Customize Spark connector endpoints for reading rows

### PR DESCRIPTION
**Pull request by anu3990**
#### Checklist: 
```diff
- Note: do not change the below
```

-  ##### Owner:

- [x] JIRA_ID included in all the commit messages
- [x] PR title is in the format JIRA_ID:Title
- [x] Rebase the branch with upstream
- [x] Squashed all commits into a single commit
- [x] Added Tests
  

- ##### Reviewer:

- [x] Reviewed Tests

### CODE REVIEW

File: marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/reader/HubDataSourceReader.java [hunk 1]
```diff
@@ -29,6 +33,11 @@
     private final Map<String, String> options;
     private final JsonNode initializationResponse;
     private final StructType sparkSchema;
+    private final HubClient hubClient;
```
Comment (rjrudin) [inline marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/reader/HubDataSourceReader.java:36]: I don't think most of these should be class fields, which complicates the class. See next note below.

File: marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/reader/HubDataSourceReader.java [hunk 2]
```diff
@@ -41,7 +50,12 @@ public HubDataSourceReader(DataSourceOptions dataSourceOptions) {
 
         this.options = dataSourceOptions.asMap();
         validateOptions(this.options);
+        this.hubClientConfig = Util.buildHubClientConfig(options);
+        this.hubClient = HubClient.withHubClientConfig(hubClientConfig);
+        this.objectMapper = new ObjectMapper();
+        this.customReadApiDefinitions = readCustomJobApiDefinitions(options);
 
+        this.endpointParams = determineReadRecordsEndpointParams(options);
```
Comment (rjrudin) [inline marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/reader/HubDataSourceReader.java:58]: I think this is all work to be done in initializeRead, which is already instantiating a HubClient. That can become a local variable in initializeRead, which can then say - if you've specified an initializereadapipath option, then I'll read that using the HubClient and pass it in as the second arg to the SparkService.on call.

File: marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/reader/HubDataSourceReader.java [hunk 3]
```diff
@@ -126,6 +145,20 @@ private int determineNumPartitions(Map<String, String> options) {
         return numPartitions;
     }
 
+    private CustomReadApiDefinitions readCustomJobApiDefinitions(Map<String, String> options) {
```
Comment (rjrudin) [inline marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/reader/HubDataSourceReader.java:148]: Because there's only one API endpoint, there's no need for a CustomReadApiDefinition class. CustomWriteApiDefinitions was created as a holder for two API definitions. We don't have that need here, so this method should just return a single InputStreamHandle. 

Also, should name this method "readCustomInitializeApiDefinition" - "Job" is not an appropriate name here.

File: marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/reader/HubDataSourceReader.java [hunk 4]
```diff
@@ -141,5 +174,49 @@ private int determineNumPartitions(Map<String, String> options) {
     public StructType readSchema() {
         return sparkSchema;
     }
+
+    private JsonNode determineReadRecordsEndpointParams(Map<String, String> options) {
```
Comment (rjrudin) [inline marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/reader/HubDataSourceReader.java:178]: Should name this "ReadRows" instead of "ReadRecords" for consistency.

File: marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/reader/HubDataSourceReader.java [hunk 5]
```diff
@@ -141,5 +174,49 @@ private int determineNumPartitions(Map<String, String> options) {
...
+        }
+
+        boolean doesNotHaveApiPath = (!endpointParams.hasNonNull("apiPath") || StringUtils.isEmpty(endpointParams.get("apiPath").asText()));
+        boolean hasEndpointConstants = endpointParams.hasNonNull("endpointConstants") && !StringUtils.isEmpty(endpointParams.get("endpointConstants").asText());
+        boolean hasEndpointState = endpointParams.hasNonNull("endpointState") && !StringUtils.isEmpty(endpointParams.get("endpointState").asText());
+        if (doesNotHaveApiPath && hasEndpointState) {
+            throw new IllegalArgumentException("Cannot set endpointState in readrowsendpointparams unless apiPath is defined as well.");
+        }
+        if(hasEndpointConstants)
+            throw new IllegalArgumentException("Cannot override endpointConstants.");
```
Comment (rjrudin) [inline marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/reader/HubDataSourceReader.java:197]: Let's give the same context as the error message above this, and I think some guidance is useful here - e.g. "Cannot set endpointConstants in readrowsendpointparams option; can only set apiPath and endpointState".

File: marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/reader/HubDataSourceReader.java [hunk 6]
```diff
@@ -141,5 +174,49 @@ private int determineNumPartitions(Map<String, String> options) {
...
+            }
+        } else {
+            endpointParams = objectMapper.createObjectNode();
+        }
+
+        boolean doesNotHaveApiPath = (!endpointParams.hasNonNull("apiPath") || StringUtils.isEmpty(endpointParams.get("apiPath").asText()));
+        boolean hasEndpointConstants = endpointParams.hasNonNull("endpointConstants") && !StringUtils.isEmpty(endpointParams.get("endpointConstants").asText());
+        boolean hasEndpointState = endpointParams.hasNonNull("endpointState") && !StringUtils.isEmpty(endpointParams.get("endpointState").asText());
+        if (doesNotHaveApiPath && hasEndpointState) {
+            throw new IllegalArgumentException("Cannot set endpointState in readrowsendpointparams unless apiPath is defined as well.");
```
Comment (rjrudin) [inline marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/reader/HubDataSourceReader.java:194]: Good error here, but add "option" after "readrowsendpointparams" to give it a little more context

File: marklogic-data-hub-spark-connector/src/test/java/com/marklogic/hub/spark/sql/sources/v2/Options.java [hunk 7]
```diff
@@ -27,6 +27,7 @@
     private String initializeWriteApiPath;
     private String finalizeWriteApiPath;
     private JsonNode additionalExternalMetadata;
+    private String initializeReadApiPath;
```
Comment (rjrudin) [inline marklogic-data-hub-spark-connector/src/test/java/com/marklogic/hub/spark/sql/sources/v2/Options.java:30]: "readRowsEndpointParams" should be here as well, and it needs to be tested too. A good way to test that, along with the custom initialize endpoint, is the following:

- Don't load any customers, nor the TDE
- Have the initialize endpoint return a single static partition, regardless of what the other inputs are
- Have the initialize endpoint return a static Spark schema as well, regardless of what the other inputs are
- For readrowsendpointparams/endpointState, define a single JSON row that can be returned by the readRows endpoint
- Then have the readRows endpoint return whatever value is in endpointState (it'll also need to use something in the endpointState object to know whether it's been called, since you only want to call it once)

That will allow you to verify that both your custom initialize and read endpoints are being used.

File: marklogic-data-hub-spark-connector/src/test/java/com/marklogic/hub/spark/sql/sources/v2/reader/ReadDataWithCustomEndpointsTest.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+
+public class ReadDataWithCustomEndpointsTest extends AbstractSparkReadTest {
+
+    @BeforeEach
+    void installCustomJobEndpoints() {
```
Comment (rjrudin) [inline marklogic-data-hub-spark-connector/src/test/java/com/marklogic/hub/spark/sql/sources/v2/reader/ReadDataWithCustomEndpointsTest.java:18]: Replace "Job" with "Read"

File: marklogic-data-hub-spark-connector/src/test/java/com/marklogic/hub/spark/sql/sources/v2/reader/ReadDataWithCustomEndpointsTest.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+        GenericDocumentManager mgr = getHubClient().getModulesClient().newDocumentManager();
+        DocumentMetadataHandle metadata = new DocumentMetadataHandle()
+            .withPermission("data-hub-operator", DocumentMetadataHandle.Capability.READ, DocumentMetadataHandle.Capability.UPDATE, DocumentMetadataHandle.Capability.EXECUTE);
+
+        String apiPath = "/custom-job-endpoints/initializeRead.api";
```
Comment (rjrudin) [inline marklogic-data-hub-spark-connector/src/test/java/com/marklogic/hub/spark/sql/sources/v2/reader/ReadDataWithCustomEndpointsTest.java:23]: Let's put these in a folder called "/custom-read-endpoints" so it's clear that they're for testing the read endpoint.

File: marklogic-data-hub-spark-connector/src/test/resources/custom-job-endpoints/initializeRead.sjs [hunk 10]
```diff
@@ -0,0 +1,121 @@
+/**
```
Comment (rjrudin) [inline marklogic-data-hub-spark-connector/src/test/resources/custom-job-endpoints/initializeRead.sjs:1]: See my note above for testing - this should return a static schema and partition object. I believe parameterizedPlan can be "{}" because your custom read endpoint won't use it.

**Review by rjrudin — CHANGES_REQUESTED** [review]

**Comment by rjrudin** [comment]
@anu3990  I'm working on a test case for this now, I'll submit a PR into your branch.

**Comment by rjrudin** [comment]
Submitted https://github.com/anu3990/marklogic-data-hub/pull/7 to provide a test that uses custom endpoints for both initializeRead and readRows.

File: marklogic-data-hub-spark-connector/src/test/java/com/marklogic/hub/spark/sql/sources/v2/reader/ReadDataWithCustomEndpointsTest.java [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+        readrowsendpointparams.put("endpointConstants", "testValue");
+        Options options = newOptions().withView("Customer").withReadRowsEndpointparams(readrowsendpointparams);
+        IllegalArgumentException ex = assertThrows(IllegalArgumentException.class,
+            () -> new HubDataSourceReader(options.toDataSourceOptions()));
+        assertEquals("Cannot set endpointConstants in readrowsendpointparams option; can only set apiPath and endpointState.",
+            ex.getMessage());
+    }
+
+    @Test
+    public void testReadRowsEndpointparams() {
```
Comment (rjrudin) [inline marklogic-data-hub-spark-connector/src/test/java/com/marklogic/hub/spark/sql/sources/v2/reader/ReadDataWithCustomEndpointsTest.java:62]: This test isn't verifying that the customReadRows.api endpoint is being used, because it's asserting that rows is empty. Rows is empty because no customers have been loaded. 

The test class I added - ReadWithCustomEndpointsTest - is failing because HubInputPartitionReader needs to look at the options to determine if it should use a custom API path or not. That test should suffice too for the happy path (you'll still want test methods for when endpointState is provided by not apiPath, and also for when endpointConstants is erroneously provided; I recommend moving both of those into ReadWithCustomEndpointsTest).

File: marklogic-data-hub-spark-connector/src/test/java/com/marklogic/hub/spark/sql/sources/v2/reader/ReadWithCustomEndpointsTest.java [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+        ObjectNode testRow = endpointState.putObject("testRow");
+        testRow.put("myId", "12345");
+        testRow.put("myName", "This is a test row");
+
+        Options options = newOptions().withView("doesnt-matter")
+            .withInitializeReadApiPath("/custom-read-endpoints/initializeRead.api")
+            .withReadRowsApiPath("/custom-read-endpoints/customReadRows.api")
+            .withReadRowsEndpointState(endpointState);
+
+        List<InternalRow> rows = readRows(new HubDataSourceReader(options.toDataSourceOptions()));
```
Comment (rjrudin) [inline marklogic-data-hub-spark-connector/src/test/java/com/marklogic/hub/spark/sql/sources/v2/reader/ReadWithCustomEndpointsTest.java:33]: Once HubInputPartitionReader is updated to use the custom endpoint, this test should pass.

File: marklogic-data-hub/src/main/resources/ml-modules/root/marklogic-data-hub-spark-connector/readRows.sjs [hunk 13]
```diff
@@ -27,4 +42,6 @@ if (endpointState.batchNumber <= partition.batchCount) {
   endpointState.batchNumber = endpointState.batchNumber + 1;
 }
 
+console.log("results", xdmp.toJsonString(results));
```
Comment (rjrudin) [inline marklogic-data-hub/src/main/resources/ml-modules/root/marklogic-data-hub-spark-connector/readRows.sjs:45]: Can you remove this line of logging? I may have accidentally included it in the PR I submitted.

File: marklogic-data-hub/src/main/resources/ml-modules/root/marklogic-data-hub-spark-connector/initializeRead.sjs [hunk 14]
```diff
@@ -20,7 +20,7 @@ const readLib = require("readLib.sjs");
 
 /**
  * Build an Optic plan based on the user's inputs.
- * 
+ *
```
Comment (rjrudin) [inline marklogic-data-hub/src/main/resources/ml-modules/root/marklogic-data-hub-spark-connector/initializeRead.sjs:23]: Undo this change just to avoid making any changes to this file (which complicates the version history of the file).

**Review by rjrudin — CHANGES_REQUESTED** [review]

File: marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/reader/HubDataSourceReader.java [hunk 15]
```diff
@@ -29,6 +33,9 @@
     private final Map<String, String> options;
     private final JsonNode initializationResponse;
     private final StructType sparkSchema;
+    private InputStreamHandle initializeDefinition;
```
Comment (rjrudin) [inline marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/reader/HubDataSourceReader.java:36]: I like to minimize the number of class fields as much as possible to simplify a class; this one can be a local variable in the initializeRead method, it doesn't need to be a class field.

File: marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/reader/HubInputPartition.java [hunk 16]
```diff
@@ -16,15 +16,17 @@
     private Map<String, String> options;
     private JsonNode initializeResponse;
     private int partitionNumber;
+    private JsonNode endpointParams;
 
-    public HubInputPartition(Map<String, String> options, JsonNode initializeResponse, int partitionNumber) {
+    public HubInputPartition(JsonNode endpointParams, Map<String, String> options, JsonNode initializeResponse, int partitionNumber) {
```
Comment (rjrudin) [inline marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/reader/HubInputPartition.java:21]: Let's make this consistent with the HubInputPartitionReader constructor, where endpointParams is the last argument out of 4.

**Review by rjrudin — CHANGES_REQUESTED** [review]
Looks good! Just a couple small tweaks to make

**Review by srinathgit — APPROVED** [review]

**Review by rjrudin — DISMISSED** [review]

**Review by rjrudin — APPROVED** [review]

**Review by srinathgit — DISMISSED** [review]

**Review by rjrudin — APPROVED** [review]

**Review by srinathgit — APPROVED** [review]
